lib/FrontendTool/FrontendTool.cpp

```diff
@@ -123,37 +130,36 @@ emitLoadedModuleTraceForAllPrimariesIfNeeded(ModuleDecl *mainModule,
       });
 }
 
-/// Gets an output stream for the provided output filename, or diagnoses to the
-/// provided AST Context and returns null if there was an error getting the
-/// stream.
-static std::unique_ptr<llvm::raw_fd_ostream>
-getFileOutputStream(StringRef OutputFilename, ASTContext &Ctx) {
-  std::error_code errorCode;
-  auto os = std::make_unique<llvm::raw_fd_ostream>(
-              OutputFilename, errorCode, llvm::sys::fs::OF_None);
-  if (errorCode) {
-    Ctx.Diags.diagnose(SourceLoc(), diag::error_opening_output,
-                       OutputFilename, errorCode.message());
-    return nullptr;
-  }
-  return os;
+/// Diagnose output file error to ASTContext.
+void diagnoseOutputError(ASTContext &Ctx, StringRef OutputFilename,
+                         llvm::Error Err) {
+  Ctx.Diags.diagnose(SourceLoc(), diag::error_opening_output, OutputFilename,
+                     toString(std::move(Err)));
 }
 
 /// Writes SIL out to the given file.
 static bool writeSIL(SILModule &SM, ModuleDecl *M, const SILOptions &Opts,
-                     StringRef OutputFilename) {
-  auto OS = getFileOutputStream(OutputFilename, M->getASTContext());
-  if (!OS) return true;
-  SM.print(*OS, M, Opts);
-
+                     StringRef OutputFilename,
+                     llvm::vfs::OutputBackend &Backend) {
+  auto OutFile = Backend.createFile(OutputFilename);
+  if (!OutFile) {
+    diagnoseOutputError(M->getASTContext(), OutputFilename,
+                        OutFile.takeError());
+    return true;
+  }
+  SM.print(*OutFile, M, Opts);
+  if (auto E = OutFile->keep()) {
```

Comment:
There is an assertion failure if you forgot to `keep()` or `discard()` the file. This is generally a good practice for all the outputs since `keep` can certainly error depending on the backend or doing an atomic rename of the file.